web/src/BulkApiButton.tsx

```diff
@@ -0,0 +1,363 @@
+  requiresConfirmation: boolean
+  targetToggleState?: ApiButtonToggleState
+  uiButtons: UIButton[]
+}
+
+type BulkApiButtonElementProps = ButtonProps & {
+  text: string
+  confirming: boolean
+  disabled: boolean
+  analyticsTags: Tags
+  analyticsName: string
+}
+
+// Styles
+const BulkButtonElementRoot = styled(InstrumentedButton)`
+  border: 1px solid ${Color.grayLight};
+  border-radius: 4px;
+  background-color: ${Color.grayLighter};
+  color: ${Color.white};
+  font-family: ${Font.monospace};
+  font-size: ${FontSize.small};
+  padding: 0 ${SizeUnit(1 / 4)};
+  text-transform: capitalize;
+  transition: color ${AnimDuration.default} ease,
+    border ${AnimDuration.default} ease;
+
+  &:hover,
+  &:active,
+  &:focus {
+    background-color: ${Color.grayLighter};
+    color: ${Color.blue};
+  }
+
+  &.Mui-disabled {
+    border-color: ${Color.grayLight};
+    color: ${Color.gray6};
+  }
+
+  /* Use shared styles with ApiButton */
+  ${confirmingButtonStateMixin}
+  ${confirmingButtonGroupBorderMixin}
+`
+
+const BulkButtonGroup = styled(ButtonGroup)<{ disabled?: boolean }>`
+  ${(props) =>
+    props.disabled &&
+    `
+    cursor: not-allowed;
+  `}
+
+  /* Adjust the border radius of buttons to achieve the "rounded bar" look */
+  &.firstButtonGroupInRow {
+    ${BulkButtonElementRoot} {
+      border-top-right-radius: 0;
+      border-bottom-right-radius: 0;
+      border-right: 0;
+    }
+  }
+
+  &.middleButtonGroupInRow {
+    ${BulkButtonElementRoot} {
+      border-radius: 0;
+      border-right: 0;
+    }
+  }
+
+  &.lastButtonGroupInRow {
+    ${BulkButtonElementRoot} {
+      border-top-left-radius: 0;
+      border-bottom-left-radius: 0;
+    }
+  }
+`
+
+// Helpers
+export function canButtonBeToggled(
+  uiButton: UIButton,
+  targetToggleState?: ApiButtonToggleState
+) {
+  const toggleInput = uiButton.spec?.inputs?.find(
+    (input) => input.name === UIBUTTON_TOGGLE_INPUT_NAME
+  )
+
+  if (!toggleInput) {
+    return false
+  }
+
+  if (!targetToggleState) {
+    return true
+  }
+
+  const toggleValue = toggleInput.hidden?.value
+
+  // A button can be toggled if it's state doesn't match the target state
+  return toggleValue !== undefined && toggleValue !== targetToggleState
+}
+
+/**
+ * A bulk button can be toggled if some UIButtons have values that don't
+ * match the target toggle state.
+ * ex: some buttons are off and target toggle is on => bulk button can be toggled
+ * ex: all buttons are on and target toggle is on   => bulk button cannot be toggled
+ * ex: all buttons are not toggle buttons           => bulk button cannot be toggled
+ */
+export function canBulkButtonBeToggled(
+  uiButtons: UIButton[],
+  targetToggleState: ApiButtonToggleState
+) {
+  const individualButtonsCanBeToggled = uiButtons.map((b) =>
+    canButtonBeToggled(b, targetToggleState)
+  )
+
+  return individualButtonsCanBeToggled.some(
+    (canBeToggled) => canBeToggled === true
+  )
+}
+
+function isBulkButtonDisabled(
```

Comment:
sure! i separated out the logic bc i thought conflating toggle-able <> disabled state may be confusing since all bulk buttons aren't toggle buttons, but it sounds like that extra layer of abstraction isn't working. i'll refactor.